app/src/main/java/com/keylesspalace/tusky/util/OkHttpUtils.kt

```diff
@@ -0,0 +1,114 @@
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version.
+ *
+ * Tusky is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
+ * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License along with Tusky. If
+ * not, see <http://www.gnu.org/licenses/>. */
+
+package com.keylesspalace.tusky.util
+
+import android.content.Context
+import android.os.Build
+import androidx.preference.PreferenceManager
+import com.keylesspalace.tusky.BuildConfig
+import okhttp3.Cache
+import okhttp3.OkHttp
+import okhttp3.OkHttpClient
+import okhttp3.tls.HandshakeCertificates
+import java.io.ByteArrayInputStream
+import java.net.InetSocketAddress
+import java.net.Proxy
+import java.security.cert.CertificateFactory
+import java.security.cert.X509Certificate
+import java.util.concurrent.TimeUnit
+
+fun okhttpClient(context: Context): OkHttpClient.Builder {
+    val preferences = PreferenceManager.getDefaultSharedPreferences(context)
+
+    val httpProxyEnabled = preferences.getBoolean("httpProxyEnabled", false)
+    val httpServer = preferences.getNonNullString("httpProxyServer", "")
+    val httpPort = preferences.getNonNullString("httpProxyPort", "-1").toIntOrNull() ?: -1
+
+    val cacheSize = 25 * 1024 * 1024 // 25 MiB
+    val builder = OkHttpClient.Builder()
+        .addInterceptor { chain ->
+            /**
+             * Add a custom User-Agent that contains Tusky, Android and Okhttp Version to all requests
+             * Example:
+             * User-Agent: Tusky/1.1.2 Android/5.0.2
+             * */
+            val requestWithUserAgent = chain.request().newBuilder()
+                .header(
+                    "User-Agent",
+                    "Tusky/${BuildConfig.VERSION_NAME} Android/${Build.VERSION.RELEASE} OkHttp/${OkHttp.VERSION}"
+                )
+                .build()
+            chain.proceed(requestWithUserAgent)
+        }
+        .readTimeout(30, TimeUnit.SECONDS)
+        .writeTimeout(30, TimeUnit.SECONDS)
+        .cache(Cache(context.cacheDir, cacheSize.toLong()))
+
+    if (httpProxyEnabled && httpServer.isNotEmpty() && httpPort > 0 && httpPort < 65535) {
+        val address = InetSocketAddress.createUnresolved(httpServer, httpPort)
+        builder.proxy(Proxy(Proxy.Type.HTTP, address))
+    }
+
+    // trust the new Let's Encrypt root certificate that is not available on Android < 7.1.1
+    // see https://letsencrypt.org/2020/11/06/own-two-feet.html
+    // see https://stackoverflow.com/questions/64844311/certpathvalidatorexception-connecting-to-a-lets-encrypt-host-on-android-m-or-ea
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+        val isgCert = """
+        -----BEGIN CERTIFICATE-----
```

Comment:
Certificate seems the same as https://letsencrypt.org/certs/isrgrootx1.pem.

Maybe also add this link for those who want to check them manually?